File: ColdDream.Api/Models/Booking.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using ColdDream.Api.Constants;

namespace ColdDream.Api.Models;

public class Booking
{
    [Key]
    public Guid Id { get; set; }

    [Required]
    public Guid UserId { get; set; }
    
    [ForeignKey(nameof(UserId))]
    public User? User { get; set; }

    [Required]
    public Guid TourRouteId { get; set; }
    
    [ForeignKey(nameof(TourRouteId))]
    public TourRoute? TourRoute { get; set; }

    public Guid? ButlerId { get; set; }
    
    [ForeignKey(nameof(ButlerId))]
    public Butler? Butler { get; set; }

    public DateTime BookingDate { get; set; }

    public int Travelers { get; set; }

    [Required]
    [MaxLength(50)]
    public string ContactName { get; set; } = string.Empty;

    [Required]
    [MaxLength(20)]
    public string ContactPhone { get; set; } = string.Empty;

    [Column(TypeName = "decimal(18,2)")]
    public decimal TotalPrice { get; set; }

    [Column(TypeName = "decimal(18,2)")]
    public decimal DiscountAmount { get; set; }

    [Required]
    [MaxLength(20)]
    public string Status { get; set; } = BookingStatus.Pending;

    // Coupon used
    public Guid? CouponId { get; set; }

    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
}

File: ColdDream.Web/src/api/booking.ts
import api, { type ApiResponse } from './index';

export interface Booking {
  id: string;
  tourRouteId: string;
  tourRoute?: any;
  bookingDate: string;
  travelers: number;
  contactName: string;
  contactPhone: string;
  totalPrice: number;
  discountAmount?: number;
  status: string;
  createdAt: string;
}

export interface CreateBookingDto {
  tourRouteId: string;
  bookingDate: string;
  travelers: number;
  contactName: string;
  contactPhone: string;
  couponId?: string;
}

export const createBooking = (data: CreateBookingDto) => {
  return api.post<any, ApiResponse<Booking>>('/booking', data);
};

export const getMyBookings = () => {
  return api.get<any, ApiResponse<Booking[]>>('/booking/my');
};

export const getBookingById = (id: string) => {
  return api.get<any, ApiResponse<Booking>>(`/booking/${id}`);
};

export const cancelBooking = (id: string) => {
  return api.post<any, ApiResponse<any>>(`/booking/${id}/cancel`);
};
